src/PBTWithTransferAuth.sol

```diff
@@ -0,0 +1,240 @@
+            uint256 tokenId = oldTokenData.tokenId;
+            _tokenDatas[newChipAddress] = TokenData(tokenId, newChipAddress, true);
+
+            if (_exists(tokenId)) {
+                emit PBTChipRemapping(tokenId, oldChipAddress, newChipAddress);
+            }
+            
+            delete _tokenDatas[oldChipAddress];
+            
+            unchecked {
+                ++i;
+            }
+        } while(i < chipAddressesOld.length);
+    }
+
+    function tokenIdFor(address chipAddress) external view override returns (uint256) {
+        uint256 tokenId = tokenIdMappedFor(chipAddress);
+        if (!_exists(tokenId)) {
+            revert NoMintedTokenForChip();
+        }
+        return tokenId;
+    }
+
+    function tokenIdMappedFor(address chipAddress) public view returns (uint256) {
+        if (!_tokenDatas[chipAddress].set) {
+            revert NoMappedTokenForChip();
+        }
+        return _tokenDatas[chipAddress].tokenId;
+    }
+
+    // Returns true if the signer of the signature of the payload is the chip for the token id
+    function isChipSignatureForToken(uint256 tokenId, bytes32 payload, bytes memory signature)
+        public
+        view
+        override
+        returns (bool)
+    {
+        if (!_exists(tokenId)) {
+            revert NoMintedTokenForChip();
+        }
+        bytes32 signedHash = keccak256(abi.encodePacked(payload)).toEthSignedMessageHash();
+        address chipAddr = signedHash.recover(signature);
+        return _tokenDatas[chipAddr].set && _tokenDatas[chipAddr].tokenId == tokenId;
+    }
+
+    //
+    // Parameters:
+    //    to: the address of the new owner
+    //    signatureFromChip: signature(receivingAddress + recentBlockhash), signed by an approved chip
+    //
+    // Contract should check that (1) recentBlockhash is a recent blockhash, (2) receivingAddress === to, and (3) the signing chip is allowlisted.
+    function _mintTokenWithChip(bytes calldata signatureFromChip, uint256 blockNumberUsedInSig)
+        internal
+        returns (uint256)
+    {
+        TokenData memory tokenData = _getTokenDataForChipSignature(signatureFromChip, blockNumberUsedInSig);
+        uint256 tokenId = tokenData.tokenId;
+        _mint(_msgSender(), tokenId);
+        emit PBTMint(tokenId, tokenData.chipAddress);
+        return tokenId;
+    }
+
+    function transferTokenWithChip(bytes calldata signatureFromChip, uint256 blockNumberUsedInSig) public override {
+        transferTokenWithChip(signatureFromChip, blockNumberUsedInSig, false);
+    }
+
+    function transferTokenWithChip(
+        bytes calldata signatureFromChip,
+        uint256 blockNumberUsedInSig,
+        bool useSafeTransferFrom
+    ) public override {
+        _transferTokenWithChip(signatureFromChip, blockNumberUsedInSig, useSafeTransferFrom);
+    }
+
+    function actionTransferRequest(uint256 _tokenId, bool _decision) public { 
+        if(msg.sender == owner) {
```

Comment:
`ownerOf(_tokenId)`

Also in the current design it seems like someone else could call `requestTransfer` to block an approved transfer.

For example, if Bob requests to buy Alice's jacket, and Alice approves, a third party Joe could overwrite Alice's approval.
We may need to model this so that each requester has their own space, similar to `_operatorApprovals` https://github.com/OpenZeppelin/openzeppelin-contracts/blob/a1948250ab8c441f6d327a65754cb20d2b1b4554/contracts/token/ERC721/ERC721.sol

Reply:
Appreciate that dude will get on the rework asap 👍🏻

Reply:
Appreciate the PR @Arkay92. Internally we want to structure these approval mechanisms as PBT extensions. I will put together a solution for "delegated receivers" which you can base your implementation off of.

Reply:
Thanks @cygaar & @2pmflow looking forward to making this happen the right way 👍🏻